lib/Twig/Cache/Filesystem.php

```diff
@@ -36,9 +36,8 @@ public function __construct($directory, $options = 0)
      */
     public function generateKey($name, $className)
     {
-        $hash = hash('sha256', $className);
-
-        return $this->directory.'/'.$hash[0].'/'.$hash[1].'/'.$hash.'.php';
+        // We assume the class name is suffixed by a hash (see Twig_Environment::getTemplateClass) that we can use to balance the directory tree.
+        return $this->directory.'/'.substr($className, -2).'/'.$className.'.php';
```

Comment:
It was done with a hash because of your comment here: https://github.com/twigphp/Twig/pull/1822#discussion_r39337173

:-1: for this change. The `getTemplateClass()` is public and can be changed in any way.